CCHMapClusterController/CCHMapClusterOperation.m

```diff
@@ -97,20 +101,237 @@ + (MKMapRect)gridMapRectForMapRect:(MKMapRect)mapRect withCellMapSize:(double)ce
+	return (a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y);
+}
+
+
++ (MKMapRect) createBoundsFromSpan:(MKMapPoint)p span:(double)span {
+	double halfSpan = span / 2;
+	return MKMapRectMake(p.x - halfSpan, p.y - halfSpan, span, span);
+}
+
+
+
 - (void)start
 {
+	switch (self.clusterMethod) {
+		case ClusterMethodGridBased:
+			[self start_grid];
+			break;
+		case ClusterMethodDistanceBased:
+			[self start_distance];
+			break;
+	};
+}
+
+
+
+// A simple clustering algorithm with O(nlog n) performance. Resulting clusters are not
+// hierarchical. Contributed by GitHub user __amishjake__.
+// Steps:
+// 1. Iterate over candidate annotations.
+// 2. Create/re-use a CCHMapClusterAnnotation with the center of the annotation.
+// 3. Add all annotations that are within a certain distance to the cluster.
+// 4. Move any annotations out of an existing cluster if they are closer to another cluster.
+// 5. Remove those items from the list of candidate clusters.
+// 6. Add/remove items on the actual map as in the grid algorithm.
+//
+- (void) start_distance
+{
+	self.executing = YES;
+	
+	BOOL respondsToSelector = [_clusterControllerDelegate respondsToSelector:@selector(mapClusterController:willReuseMapClusterAnnotation:)];
+
+	double zoomLevel = CCHMapClusterControllerZoomLevelForRegion(self.mapViewRegion.center.longitude, self.mapViewRegion.span.longitudeDelta, self.mapViewWidth);
+	BOOL disableClustering = (zoomLevel > self.maxZoomLevelForClustering);
+
+	// Zoom scale * MK distance = screen points
+	MKZoomScale zoomScale = [self currentZoomScale];
+	
+	// The width and height of the square around a point that we'll consider later
+	double zoomSpecificSpan = _cellSize / zoomScale;
+	
+	// Annotations we've already looked at for a starting point for a cluster
+	NSMutableSet *visitedCandidates = [[NSMutableSet alloc] init];
+
+	// The MKAnnotations (single POIs and clusters alike) we want on display
+	NSMutableSet *results = [[NSMutableSet alloc] init];
+	
+	// Map a single POI MKAnnotation to its distance (NSNumber*) from its cluster (if added to one yet)
+	NSMapTable *distanceToCluster = [NSMapTable strongToStrongObjectsMapTable];
+	
+	// Map a single POI MKAnnotation to its cluster annotation (if added to one yet)
+	NSMapTable *itemToCluster = [NSMapTable strongToStrongObjectsMapTable];
+
+	// Existing annotations to re-use
+	NSMutableSet *annotationsToReuse = [NSMutableSet setWithSet:[_visibleAnnotationsMapTree annotationsInMapRect:self.mapView.visibleMapRect]];
+	//NSMutableSet *annotationsToReuse = [NSMutableSet setWithSet:[_visibleAnnotationsMapTree annotations]];
+	
+	// The CCHMapClusterAnnotations that are being reused
+	NSMutableSet *reusedAnnotations = [NSMutableSet setWithSet:[_visibleAnnotationsMapTree annotations]];
+	
+	@synchronized(_allAnnotationsMapTree) {
+		
+		// Iterate over all POIs we know about
+		for (id<MKAnnotation> candidate in _allAnnotationsMapTree.annotations) {
```

Comment:
This might not be the most performant way. Imagine we have tons of annotations placed over the whole globe. We'd iterate over all of them when ever we change the visible map rect for a small area. In the grid based algorithm we only consider annotations in a specific area around the visible map rect to speed up performance. I guess here we could do the same?